### PR TITLE
[Feature] @s-ui/studio run tests in the cli

### PR DESCRIPTION
**Pull request by carlosvillu**
Using a real DOM.

Captura de pantalla 2020-05-22 a las 14 46 14

runtime.js
```js
import React from 'react'
import hoistNonReactStatics from 'hoist-non-react-statics'

import SUIContext from '@s-ui/react-context'
import withContext from '@s-ui/studio/src/components/demo/HoC/withContext'
import {cleanDisplayName} from '@s-ui/studio/src/components/demo/utilities'

import AtomButton from '../../../components/atom/button/src'
import AtomButtonContexts from '../../../demo/atom/button/context'

const originalKarmaLoader = window.__karma__.loaded
window.__karma__.loaded = () => {}

const run = async () => {
  const nextAtomButtonContexts =
    typeof AtomButtonContexts !== 'function'
      ? AtomButtonContexts
      : await AtomButtonContexts()

  debugger // eslint-disable-line
  window.__STUDIO_CONTEXTS__ = nextAtomButtonContexts
  window.__STUDIO_COMPONENT__ = AtomButton

  const EnhanceComponent = withContext(
    nextAtomButtonContexts.default,
    nextAtomButtonContexts
  )(AtomButton)
  !EnhanceComponent.displayName &&
    console.error('[sui-Test] Component without displayName') // eslint-disable-line

  const NextComponent = props => (
    <SUIContext.Provider value={nextAtomButtonContexts.default}>
      <EnhanceComponent {...props} />
    </SUIContext.Provider>
  )
  hoistNonReactStatics(NextComponent, AtomButton)

  const displayName = cleanDisplayName(EnhanceComponent.displayName)
  NextComponent.displayName = displayName
  window[displayName] = NextComponent

  await import('./index.js')
  originalKarmaLoader.call(window.__karma__)
}

run()
```

---
This change is [Reviewable](https://reviewable.io/reviews/sui-components/sui/834)

### CODE REVIEW

File: packages/sui-studio/src/patcher-mocha/index.js [hunk 1]
```diff
@@ -6,6 +6,7 @@ import withContext from '../components/demo/HoC/withContext'
 import {cleanDisplayName} from '../components/demo/utilities'
 import hoistNonReactStatics from 'hoist-non-react-statics'
 
+debugger // eslint-disable-line
```
Comment (arnau-rius) [inline packages/sui-studio/src/patcher-mocha/index.js:9]: 👀

Comment (andresin87) [reply packages/sui-studio/src/patcher-mocha/index.js:9]: WIP

**Comment by carlosvillu** [comment]
Runtime for several components:

```js
import React from 'react'
import hoistNonReactStatics from 'hoist-non-react-statics'

import SUIContext from '@s-ui/react-context'
import withContext from '@s-ui/studio/src/components/demo/HoC/withContext'
import {cleanDisplayName} from '@s-ui/studio/src/components/demo/utilities'

import AtomButton from '../components/atom/button/src'
import AtomButtonContexts from '../demo/atom/button/context'

import AtomHeader from '../components/atom/header/src'
import AtomHeaderContexts from '../demo/atom/header/context'

const originalKarmaLoader = window.__karma__.loaded
window.__karma__.loaded = () => {}

const enhanceComponent = displayName => {
  !displayName && console.error('[sui-Test] Component without displayName') // eslint-disable-line

  const activeContext = window.__STUDIO_CONTEXTS__[displayName]
  const ActiveComponent = window.__STUDIO_COMPONENT__[displayName]

  const EnhanceComponent = withContext(
    activeContext.default,
    activeContext
  )(ActiveComponent)

  const NextComponent = props => (
    <SUIContext.Provider value={activeContext.default}>
      <EnhanceComponent {...props} />
    </SUIContext.Provider>
  )
  hoistNonReactStatics(NextComponent, ActiveComponent)

  const nextDisplayName = cleanDisplayName(displayName)
  NextComponent.displayName = nextDisplayName

  return NextComponent
}

const run = async () => {
  const nextAtomButtonContexts = typeof AtomButtonContexts !== 'function' ? AtomButtonContexts : await AtomButtonContexts() // eslint-disable-line
  const nextAtomHeaderContexts = typeof AtomHeaderContexts !== 'function' ? AtomHeaderContexts : await AtomHeaderContexts() // eslint-disable-line

  window.__STUDIO_CONTEXTS__ = {}
  window.__STUDIO_CONTEXTS__[AtomButton.displayName] = nextAtomButtonContexts
  window.__STUDIO_CONTEXTS__[AtomHeader.displayName] = nextAtomHeaderContexts

  window.__STUDIO_COMPONENT__ = {}
  window.__STUDIO_COMPONENT__[AtomButton.displayName] = AtomButton
  window.__STUDIO_COMPONENT__[AtomHeader.displayName] = AtomHeader

  window[AtomButton.displayName] = enhanceComponent(AtomButton.displayName) // eslint-disable-line
  window[AtomHeader.displayName] = enhanceComponent(AtomHeader.displayName) // eslint-disable-line

  await import('./atom/button/index.js')
  await import('./atom/header/index.js')
  originalKarmaLoader.call(window.__karma__)
}

run()
```

https://github.com/SUI-Components/sui/blob/master/packages/sui-dashboard/bin/sui-dashboard-components.js#L19


Captura de pantalla 2020-05-22 a las 14 46 14

**Comment by andresin87** [comment]
```js
import React from 'react'
import hoistNonReactStatics from 'hoist-non-react-statics'

import SUIContext from './Context'
import withContext from '@s-ui/studio/src/components/demo/HoC/withContext'
import {cleanDisplayName} from '@s-ui/studio/src/components/demo/utilities'

const requireFile = async ({
  defaultValue,
  extractDefault = true,
  importFile
}) => {
  const file = await importFile().catch(_ => defaultValue)
  if (typeof file === 'undefined') {
    return Promise.reject(new Error('Error requiring file'))
  }
  return extractDefault ? file.default : file
}

const tryRequireContext = ({category, component}) =>
  requireFile({
    defaultValue: false,
    importFile: () => {
      return import(`../demo/${category}/${component}/context.js`)
    }
  })

const tryRequireComponent = ({category, component}) =>
  requireFile({
    defaultValue: false,
    importFile: () =>
      import(`../components/${category}/${component}/src/index.js`)
  })

const regex = /\.\/(?<categoryName>\w+)\/(?<componentName>\w+)\/(src)+\/(index.js|jsx)$/

function declareAll(resolve) {
  const cache = {}
  const capitalize = s => {
    if (typeof s !== 'string') return ''
    return s.charAt(0).toUpperCase() + s.slice(1)
  }
  resolve.keys().forEach(srcKey => {
    const {
      groups: {categoryName, componentName}
    } = srcKey.match(regex)
    const component = resolve(srcKey).default
    const key =
      component.displayName ||
      capitalize(categoryName) + capitalize(componentName)
    cache[key] = {
      displayName: key,
      categoryName,
      componentName
    }
  })
  return cache
}
function testAll(resolve) {
  resolve.keys().forEach(key => {
    resolve(key)
  })
}

const originalKarmaLoader = window.__karma__.loaded
window.__karma__.loaded = () => {}

const enhanceComponent = displayName => {
  !displayName && console.error('[sui-Test] Component without displayName') // eslint-disable-line

  const activeContext = window.__STUDIO_CONTEXTS__[displayName]
  const ActiveComponent = window.__STUDIO_COMPONENT__[displayName]

  const EnhanceComponent = withContext(
    activeContext.default,
    activeContext
  )(ActiveComponent)

  const NextComponent = props => (
    <SUIContext.Provider value={activeContext.default}>
      <EnhanceComponent {...props} />
    </SUIContext.Provider>
  )
  hoistNonReactStatics(NextComponent, ActiveComponent)

  const nextDisplayName = cleanDisplayName(displayName)
  NextComponent.displayName = nextDisplayName

  return NextComponent
}

const run = async () => {
  const cache = declareAll(
    require.context(
      '../components/',
      true,
      /\.\/(\w+)\/(\w+)\/(src)+\/index.(js|jsx)$/
    )
  )

  async function asyncForEach(array, callback) {
    const response = []
    for (let index = 0; index < array.length; index++) {
      response[index] = await callback(array[index], index, array)
    }
    return response
  }

  window.__STUDIO_CONTEXTS__ = {}
  window.__STUDIO_COMPONENT__ = {}
  await asyncForEach(
    Object.keys(cache),
    async cachedValueKeyName => {
      const {categoryName, componentName, displayName} = cache[
        cachedValueKeyName
      ]
      let nextEntityContexts = await tryRequireContext({
        category: categoryName,
        component: componentName
      })
      nextEntityContexts =
        typeof nextEntityContexts !== 'function'
          ? nextEntityContexts
          : await nextEntityContexts() // eslint-disable-line
      const nextEntityComponent = await tryRequireComponent({
        category: categoryName,
        component: componentName
      })
      window.__STUDIO_CONTEXTS__[displayName] = nextEntityContexts
      window.__STUDIO_COMPONENT__[displayName] = nextEntityComponent
      window[displayName] = enhanceComponent(displayName) // eslint-disable-line
      return {
        displayName,
        categoryName,
        componentName,
        nextEntityContexts,
        nextEntityComponent
      }
    }
  )

  testAll(
    require.context('../test/', true, /\.\/(\w+)\/(\w+)\/index.spec.(js|jsx)$/)
  )

  originalKarmaLoader.call(window.__karma__)
}

run()

```

File: packages/sui-test/bin/karma/babelPatch.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
+module.exports = function (babel) {
```
Comment (andresin87) [inline packages/sui-test/bin/karma/babelPatch.js:1]: maybe this file might be in sui-codemods renamed as `sui-test-describe-context-patch`?

**Comment by andresin87** [comment]
Babelized codemod produced code is already cached in `nodemodules/.cache`. If we want to modify the codemod its important to remove this folder in order to get the desired actual result

**Comment by andresin87** [comment]
# UPDATE
## Purpose
Execute all the test of a project using CLI for CI/CD intentions.

File: packages/sui-studio/src/patcher-mocha/index.js [hunk 3]
```diff
@@ -24,43 +23,53 @@ Function.prototype.partial = function() {
...
-          <EnhanceComponentWithLegacyContext {...props} />
-        </SUIContext.Provider>
-      )
-      hoistNonReactStatics(EnhanceComponent, EnhanceComponentWithLegacyContext)
+        const EnhanceComponentWithLegacyContext = withContext(
+          context,
+          context
+        )(__COMPONENT__)
+        const EnhanceComponent = props => {
+          debugger // eslint-disable-line
```
Comment (jcger) [inline packages/sui-studio/src/patcher-mocha/index.js:61]: 👀

Comment (andresin87) [reply packages/sui-studio/src/patcher-mocha/index.js]: done

File: packages/sui-studio/bin/sui-studio-test.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+/* eslint-disable no-console */
+
+const program = require('commander')
+const {serialSpawn} = require('@s-ui/helpers/cli')
+const path = require('path')
+
+const suiTestClientPath = require.resolve('@s-ui/test/bin/sui-test-browser')
+
+program
+  .option('--ci', 'Run compoents tests in CLI, headless mode')
```
Comment (jcger) [inline packages/sui-studio/bin/sui-studio-test.js:11]: typo compoents instead of components

File: packages/sui-studio/src/runtime-mocha/index.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+import SUIContext from '@s-ui/react-context'
+import withContext from '../components/demo/HoC/withContext'
+import {cleanDisplayName} from '../components/demo/utilities'
+import {requireFile, tryRequireContext, tryRequireComponent} from '../components/tryRequire'
+
+const originalKarmaLoader = window.__karma__.loaded
+window.__karma__.loaded = () => {}
+
+const regex = /\.\/(?<categoryName>\w+)\/(?<componentName>\w+)\/(src)+\/(index.js|jsx)$/
```
Comment (jcger) [inline packages/sui-studio/src/runtime-mocha/index.js:13]: I guess this will break if the mono repo depth is 3 instead of 2, right?

Comment (andresin87) [reply packages/sui-studio/src/runtime-mocha/index.js:13]: yeap
[https://regexr.com/570ub](https://regexr.com/570ub)

File: packages/sui-test/bin/karma/index.js [hunk 6]
```diff
@@ -39,10 +39,12 @@ module.exports = async ({
 
   configure.files = [
     `${CWD}/node_modules/@babel/polyfill/dist/polyfill.min.js`,
+    srcPattern && `${CWD}/${srcPattern}`,
```
Comment (rmoralp) [inline packages/sui-test/bin/karma/index.js:42]: Following code added in L47 https://github.com/SUI-Components/sui/pull/834/files#diff-3259a4ee8eb62198bf98bcc989d16990R47

You can avoid filter undefined by doing this:
```suggestion
    ...(srcPattern && [`${CWD}/${srcPattern}`]),
```

File: packages/sui-studio/src/runtime-mocha/index.js (hunk 5 above)
Comment (carlosvillu) [reply packages/sui-studio/src/runtime-mocha/index.js:13]: Well, this really isn't for any monorepo, it's just for the Studios, which we already know are only two levels deep.

File: packages/sui-studio/src/runtime-mocha/index.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    })
+    nextEntityContexts =
+      typeof nextEntityContexts !== 'function'
+        ? nextEntityContexts
+        : await nextEntityContexts() // eslint-disable-line
+    const nextEntityComponent = await tryRequireComponent({
+      category: categoryName,
+      component: componentName
+    })
+    window.__STUDIO_CONTEXTS__[displayName] = nextEntityContexts
```
Comment (rmoralp) [inline packages/sui-studio/src/runtime-mocha/index.js:98]: While development, sometimes we wrap a component in `React.memo` which removes `displayName`.
Is it possible to make a warning if displayName does not exist?

File: packages/sui-studio/src/patcher-mocha/index.js [hunk 8]
```diff
@@ -24,43 +23,52 @@ Function.prototype.partial = function() {
...
+          global.__STUDIO_COMPONENT__
+
+        const __CONTEXTS__ =
+          global.__STUDIO_CONTEXTS__[__COMPONENT__.displayName] ||
+          global.__STUDIO_CONTEXTS__ ||
+          {}
+
+        if (!__CONTEXTS__) {
+          console.error(
+            `You're trying to use the context ${prop} but it's not defined in your contexts.js file`
```
Comment (javiauso) [inline packages/sui-studio/src/patcher-mocha/index.js:41]: contexts.js -> context.js

File: packages/sui-studio/src/patcher-mocha/index.js [hunk 9]
```diff
@@ -24,43 +23,52 @@ Function.prototype.partial = function() {
...
-      let context = __CONTEXTS__[prop]
-      if (!context) {
-        // eslint-disable-next-line
-        console.error(
-          `Your trying to use the context ${prop} but it is not define in your contexts.js file.
+        let context = __CONTEXTS__[prop]
+        if (!context) {
+          // eslint-disable-next-line
+          console.error(
+            `Your trying to use the context ${prop} but it is not defined in your contexts.js file.
```
Comment (javiauso) [inline packages/sui-studio/src/patcher-mocha/index.js:50]: contexts.js -> context.js

File: packages/sui-studio/src/runtime-mocha/index.js (hunk 7 above)
Comment (andresin87) [reply packages/sui-studio/src/runtime-mocha/index.js:98]: we realized that and fixed it using
`const displayName = capitalize(categoryName) + capitalize(componentName)`
line 26

Comment (andresin87) [reply packages/sui-studio/src/runtime-mocha/index.js:98]: the display name is provided by the babelPatcher AST plugin

**Pull request merged**